### PR TITLE
Add some improvements for pipeline module, engine and assertion into ds engine

### PR DESCRIPTION
**Pull request by hyunwoongko**
There are some improvements.

1. Improve error message of pipeline engine. see https://github.com/microsoft/DeepSpeed/pull/1438
2. Move `has_bool_tensors` to `kwargs` to make usable with `deepspeed.initialize` and rename `has` to `send`. see https://github.com/microsoft/DeepSpeed/pull/1399
3. Add assertion to DS engine. see https://github.com/microsoft/DeepSpeed/issues/1034
4. Add `@property` to make `module.mpu` usable.
    - when we use `PipelineEngine`, we use syntax like `engine.mpu`. but when we use `PipelineModule` we must use `module.mpu()`. this can be confusing front end users.
    - So, I added `@property` to make `module.mpu` usable. We can use same syntax like `something.mpu`.
    - I also added `__call__()` method to `PipelineParallelGrid` to be compatible with exsting code

cc. @tjruwase

### CODE REVIEW

**Comment by hyunwoongko** [comment]
I can't understand why formatting CI was failed... 😂

스크린샷 2021-11-06 오후 1 44 11

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -689,6 +689,8 @@ def _do_args_sanity_check(self, args):
                 assert args.deepspeed_config is None, "Not sure how to proceed, we were given both a deepscale_config and deepspeed_config"
             args.deepspeed_config = args.deepscale_config
 
+        assert self.client_optimizer is not None or self.client_model_parameters is not None, "At least one of the optimizer and model parameters must be input to train the model."
```
Comment (hyunwoongko) [inline deepspeed/runtime/engine.py:692]: 1. Add assertion to DS engine. see https://github.com/microsoft/DeepSpeed/issues/1034

File: deepspeed/runtime/pipe/engine.py [hunk 2]
```diff
@@ -1207,16 +1207,19 @@ def _allocate_buffers(self, shapes_and_dtypes, requires_grad=False, num_buffers=
         return buffers
 
     def forward(self, *args, **kwargs):
-        """Disabled for pipeline parallel training. See ``train_batch()``. """
-        raise PipelineError("Only train_batch() is accessible in pipeline mode.")
+        """Disabled for pipeline parallel training. See ``train_batch()`` and ``eval_batch()``."""
```
Comment (hyunwoongko) [inline deepspeed/runtime/pipe/engine.py:1210]: 2. Improve error message of pipeline engine. see https://github.com/microsoft/DeepSpeed/pull/1438

File: deepspeed/runtime/pipe/module.py [hunk 3]
```diff
@@ -512,6 +512,7 @@ def topology(self):
         """ ProcessTopology object to query process mappings. """
         return self._topo
 
+    @property
```
Comment (hyunwoongko) [inline deepspeed/runtime/pipe/module.py:515]: 3. Add `@property` to make `pipe_module.mpu` usable.

File: deepspeed/runtime/pipe/engine.py [hunk 4]
```diff
@@ -65,15 +65,15 @@ class PipelineEngine(DeepSpeedEngine):
...
+    def __init__(self, *super_args, **super_kwargs):
         super().__init__(*super_args, **super_kwargs)
         assert isinstance(self.module, PipelineModule), "model must base PipelineModule"
 
         assert self.zero_optimization_stage() < 2, "ZeRO-2 and ZeRO-3 are incompatible with pipeline parallelism"
 
         # We schedule the all-reduces, so disable it in super().backward()
         self.enable_backward_allreduce = False
-        self.has_bool_tensors = has_bool_tensors
+        self.has_bool_tensors = super_kwargs.get("send_bool_tensors", False)
```
Comment (hyunwoongko) [inline deepspeed/runtime/pipe/engine.py:76]: 4. Move `has_bool_tensors` to `kwargs` to make usable with `deepspeed.initialize` and rename `has` to `send`. see https://github.com/microsoft/DeepSpeed/pull/1399

File: deepspeed/runtime/engine.py [hunk 5]
```diff
@@ -721,6 +721,9 @@ def _do_sanity_check(self):
                 assert self._is_supported_optimizer(self.optimizer_name()), \
                     '{} is not a supported DeepSpeed Optimizer'.format(self.optimizer_name())
 
+        assert self.client_optimizer is not None or self.client_model_parameters is not None, \
```
Comment (hyunwoongko) [inline deepspeed/runtime/engine.py:724]: 1. Add assertion to DS engine. see https://github.com/microsoft/DeepSpeed/issues/1034

**Comment by hyunwoongko** [comment]
```
AssertionError: At least one of the `optimizer` and `model_parameters` must be input to train the model.
```
we should modify test code

**Comment by rocm-mici** [comment]
Can one of the admins verify this patch?

**Comment by loadams** [comment]
Hi @hyunwoongko - apologies for taking so long to reply to this.  Is this still a branch you'd like to see merged?  If so, could you resolve merge conflicts?  If not, we're closing and cleaning up old branches in the repo.

**Pull request closed without merging**